File: packages/apple/AvatarAnimationEngine/Sources/Scene/Rig.swift
import Foundation

/// One paintable thing, fully resolved: where it is, what it looks like, and how
/// it is painted. The engine produces an array of these per frame and hands it
/// to a renderer that knows nothing about rigs, channels or config.
public struct DisplayItem: Equatable, Sendable {
    public struct Paint: Equatable, Sendable {
        /// Always a `#rrggbb` literal — palette keys and `"@node"` late bindings
        /// have already been resolved away, so no renderer needs the palette.
        public var ink: String
        public var alpha: Double
        public var fill: Bool
        public var width: Double?

        public init(ink: String, alpha: Double, fill: Bool, width: Double?) {
            self.ink = ink; self.alpha = alpha; self.fill = fill; self.width = width
        }
    }

    public var id: String
    /// World transform, in design units.
    public var m: Mat
    /// The resolved path, in the node's LOCAL space. The matrix carries the
    /// placement; the points are never baked.
    public var d: String
    /// The parsed path's command string, e.g. `"MCCCCZ"`.
    public var kind: String
    public var paint: Paint

    public init(id: String, m: Mat, d: String, kind: String, paint: Paint) {
        self.id = id; self.m = m; self.d = d; self.kind = kind; self.paint = paint
    }
}

public typealias DisplayList = [DisplayItem]

public struct SceneError: Error, Equatable, CustomStringConvertible {
    public let message: String
    public init(_ message: String) { self.message = message }
    public var description: String { "avatar scene: \(message)" }
}

/// A flattened, variant-folded rig, ready to compose. Building one is the only
/// place a variant is ever mentioned: everything downstream — compose, the tween
/// engine, the golden recorder, the renderer — is unaware variants exist.
public struct Scene {
    /// A node plus the id of the node whose world transform it inherits.
    ///
    /// The web keeps the whole ancestor chain and reads only its last element.
    /// One parent id is the same information, and it says out loud that a node's
    /// placement depends on exactly one other node.
    struct Placed {
        let node: RigNode
        let parent: String?
    }

    public let config: CharacterConfig
    /// Depth-first, declaration order — the order the display list will carry.
    let flat: [Placed]
    /// `config.character.inks` with this scene's variant folded in, identical to
    /// it when there is none. `compose` reads paint from HERE and never from the
    /// config, which is what keeps a variant from leaking anywhere else.
    let inks: [String: Ink]

    /// `variant` names a `character.variants` entry and is applied once, here.
    /// The loader has already guaranteed each patch names a real ink whose fields
    /// exist, or a real node that has no `.shape` channel and whose named fields
    /// exist with the same type and arity — so the folding below has nothing left
    /// to validate. An unknown name throws; `nil` means the true rig.
    public init(_ config: CharacterConfig, variant: String? = nil) throws {
        var patch: Variant?
        if let variant {
            guard let found = config.character.variants[variant] else {
                throw SceneError("unknown variant \"\(variant)\"")
            }
            patch = found
        }

        var flat: [Placed] = []
        func walk(_ node: RigNode, _ parent: String?) {
            var node = node
            if var shape = node.shape, let fields = patch?.shapes?[node.id] {
                for (name, value) in fields.sorted(by: { $0.key < $1.key }) {
                    shape.patch(name, value)
                }
                node.shape = shape
            }
            flat.append(Placed(node: node, parent: parent))
            for child in node.children ?? [] { walk(child, node.id) }
        }
        walk(config.rig.root, nil)

        var inks = config.character.inks
        for (name, over) in (patch?.inks ?? [:]).sorted(by: { $0.key < $1.key }) {
            guard var ink = inks[name] else { continue }
            for (field, value) in over.sorted(by: { $0.key < $1.key }) {
                switch (field, value) {
                case ("kind", .string(let v)): ink.kind = Ink.Kind(rawValue: v) ?? ink.kind
                case ("color", .string(let v)): ink.color = v
                case ("width", .number(let v)): ink.width = v
                default: break     // unreachable: the loader rejected it
                }
            }
            inks[name] = ink
        }

        self.config = config
        self.flat = flat
        self.inks = inks
    }

    public func compose(_ channels: Channels) throws -> DisplayList {
        // Written by id, read by id, never iterated — so a Dictionary is safe
        // here in a way it is not anywhere else in this engine.
        var world: [String: Mat] = [:]
        var out: DisplayList = []
        out.reserveCapacity(flat.count)

        for placed in flat {
            let node = placed.node
            let id = node.id
            let parent = placed.parent.flatMap { world[$0] } ?? .identity
            let local = Mat.from(
                x: num(channels, "\(id).x", 0),
                y: num(channels, "\(id).y", 0),
                rotation: num(channels, "\(id).rotation", 0),
                scaleX: num(channels, "\(id).scaleX", 1),
                scaleY: num(channels, "\(id).scaleY", 1),
                // Channels, not `node.transform.pivot`: the rig only seeds the
                // rest value. The original moves a transform origin as freely as
                // it moves a rotation — the sad droop turns the face about its
                // bbox bottom, the settle turns it about 60% height — so the
                // origin has to be animatable too.
                pivot: (num(channels, "\(id).pivotX", 0),
                        num(channels, "\(id).pivotY", 0)))
            let m = parent * local
            world[id] = m

            // A node with no shape still transforms its children — `body` is
            // exactly that, and it is where the whole character's placement and
            // its late-bound colour live.
            guard let shape = node.shape else { continue }

            let d = try resolvePath(id: id, shape: shape, channels: channels)
            // The channel is authoritative when present: an ink can be swapped
            // at runtime, and the loader seeded this from the rig.
            let inkName = channels.get("\(id).ink")?.text ?? node.ink ?? ""
            // Stroke-vs-fill and the stroke width are properties of the INK, not
            // of the shape. That is what lets one ink restyle six nodes from one
            // place, and it is the single fact CoreGraphics has to agree with SVG
            // about.
            let ink = inks[inkName]
            out.append(DisplayItem(
                id: id,
                m: m,
                d: d,
                kind: try parsePath(d).kind,
                paint: DisplayItem.Paint(
                    ink: try resolveInk(inkName, channels),
                    alpha: num(channels, "\(id).alpha", 1),
                    fill: ink?.kind == .fill,
                    width: ink?.width)))
        }
        return out
    }

    /// Rebuild a node's path.
    ///
    /// A morphable node's geometry lives on its `.shape` channel — seeded from
    /// the rig, then driven by poses, timelines and morphs — so the channel is
    /// authoritative whenever the node has one. The primitives have no `.shape`
    /// channel and are rebuilt from the rig each frame, because their own
    /// parameters and `bend` are what animate them.
    private func resolvePath(id: String, shape: Shape, channels: Channels) throws -> String {
        if let driven = channels.get("\(id).shape")?.text { return driven }
        // `Build` speaks flat number pairs — the Path module knows nothing
        // about config types, so `Pt` is unpacked here.
        let pairs = { (points: [Pt]) -> [[Double]] in points.map { [$0.x, $0.y] } }
        switch shape.kind {
        case .ring:
            return Build.ring(cx: shape.cx!, cy: shape.cy!, r: shape.r!, band: shape.band!)
        case .disc:
            return Build.disc(cx: shape.cx!, cy: shape.cy!, r: shape.r!)
        case .arc:
            return Build.arc(cx: shape.cx!, cy: shape.cy!, r: shape.r!,
                             from: shape.from!, to: shape.to!)
        case .cubicO:
            return Build.cubicO(cx: shape.cx!, cy: shape.cy!, rx: shape.rx!, ry: shape.ry!)
        case .polyline:
            return try Build.polyline(pairs(shape.points!))
        case .bezier:
            guard let bend = shape.bend else { return try Build.bezier(pairs(shape.points!)) }
            // `bend` offsets each control point along one axis by its weight.
            // The channel ALREADY carries the inward damp —
            // `CharacterConfig.respond` applies it where the value is written,
            // because the original damps a tween's endpoints and lerps the
            // paths between them. Damping again here would bend the line a
            // second time, and the kink would land mid-stroke.
            let a = num(channels, "\(id).bend", 0)
            let axis = bend.axis == "x" ? 0 : 1
            var moved = pairs(shape.points!)
            for i in moved.indices {
                moved[i][axis] += (i < bend.weights.count ? bend.weights[i] : 0) * a
            }
            return try Build.bezier(moved)
        }
    }

    /// ink key | palette key | `"@nodeId"` | `#rrggbb` -> a literal colour.
    ///
    /// `"@body"` is late-bound: one channel recolours the whole character, so the
    /// rig never repeats a mood colour on fourteen nodes. The depth guard is not
    /// decoration — a config that made `"@x"` resolve back to itself would
    /// otherwise hang the render loop rather than fail, and the loader cannot see
    /// a cycle that only closes through a runtime channel value.
    private func resolveInk(_ raw: String, _ channels: Channels, depth: Int = 0) throws -> String {
        guard depth <= 8 else {
            throw SceneError("ink \"\(raw)\" does not resolve to a colour")
        }
        // A colour channel already holds a literal: the loader normalised the
        // resting value, and a tween mid-flight produces a new one every frame.
        if raw.hasPrefix("#") { return raw }
        if raw.hasPrefix("@") { return try resolveLateBound(raw, channels, depth: depth) }
        if let ink = inks[raw] {
            return try resolveColourRef(ink.color, channels, depth: depth + 1)
        }
        guard let colour = config.character.palette[raw] else {
            throw SceneError("unknown colour \"\(raw)\"")
        }
        return colour
    }

    /// Resolves an `Ink.color` field. By that field's own contract it is always
    /// `"#hex"`, `"@nodeId"`, or a bare PALETTE key -- never another ink key --
    /// so, unlike `resolveInk`, this never re-enters the ink table. That
    /// distinction is load-bearing on real data, not pedantry: an ink is
    /// routinely named after the very palette colour it points at, and recursing
    /// through the ink table would find that same-named ink again and resolve it
    /// to itself. On the web that spins; here the depth guard above catches it
    /// and throws instead -- which is worse to diagnose, not better, because the
    /// character simply refuses to paint and the config it was handed is valid.
    private func resolveColourRef(_ raw: String, _ channels: Channels, depth: Int) throws -> String {
        guard depth <= 8 else {
            throw SceneError("ink \"\(raw)\" does not resolve to a colour")
        }
        if raw.hasPrefix("#") { return raw }
        if raw.hasPrefix("@") { return try resolveLateBound(raw, channels, depth: depth) }
        guard let colour = config.character.palette[raw] else {
            throw SceneError("unknown colour \"\(raw)\"")
        }
        return colour
    }

    /// A `.ink` channel's value may name an ink, so this re-enters the full lookup.
    private func resolveLateBound(_ raw: String, _ channels: Channels, depth: Int) throws -> String {
        let node = String(raw.dropFirst())
        guard let driven = channels.get("\(node).ink")?.text else {
            throw SceneError("late-bound ink \"\(raw)\" has no value on \(node).ink")
        }
        return try resolveInk(driven, channels, depth: depth + 1)
    }
}

extension CharacterConfig {
    /// Write every channel's rest value into the store. The loader already
    /// derived the map from the rig tree (Task 29), so there is exactly one
    /// definition of "at rest" and this cannot drift from it.
    public func seed(into channels: Channels) {
        for name in rest.keys.sorted() { channels.set(name, rest[name]!) }
    }
}

private func num(_ channels: Channels, _ name: String, _ fallback: Double) -> Double {
    channels.get(name)?.number ?? fallback
}

File: packages/web/packages/avatar-engine/src/scene/rig.ts
import { IDENTITY, fromTransform, multiply, type Mat } from "../math/mat";
import { arc, bezier, cubicO, disc, polyline, ring } from "../path/build";
import { parsePath } from "../path/parse";
import type { Channels, ChannelValue } from "../runtime/channels";
import type { CharacterConfig, Ink, RigNode } from "../config/types";

export interface DisplayItem {
  id: string;
  m: Mat;
  d: string;
  kind: string;
  paint: { ink: string; alpha: number; fill: boolean; width?: number };
}
export type DisplayList = readonly DisplayItem[];

export interface Scene {
  config: CharacterConfig;
  flat: readonly { node: RigNode; parents: readonly string[] }[];
  /** `config.character.inks` with this scene's variant folded in — identical to
   *  it when there is no variant. Compose reads paint from here and never from
   *  the config, which is what keeps a variant from leaking anywhere else. */
  inks: Readonly<Record<string, Ink>>;
}

/** Depth-first, declaration order — the order the display list will carry.
 *
 *  `variant` names a `character.variants` entry (Task 1) and is applied HERE,
 *  once, so that everything downstream — compose, the tween engine, the golden
 *  recorder — is unaware variants exist. The loader has already guaranteed each
 *  patch names a real ink, or a real node that has no `.shape` channel and whose
 *  named fields exist with the same type and arity. */
export function buildScene(config: CharacterConfig, variant?: string): Scene {
  // `hasOwn`, not a plain index-then-`undefined` check: `variants` is a
  // JSON-derived Record, so a plain index sees `Object.prototype` too, and
  // `variants["constructor"]` comes back truthy without ever naming a real
  // variant — exactly the silent-true-rig outcome this guard exists to forbid.
  if (variant !== undefined && !Object.hasOwn(config.character.variants, variant)) {
    throw new Error(`avatar: unknown variant "${variant}"`);
  }
  const patch = variant === undefined ? undefined : config.character.variants[variant];

  const flat: { node: RigNode; parents: string[] }[] = [];
  const walk = (node: RigNode, parents: string[]): void => {
    const shapePatch = patch?.shapes?.[node.id];
    flat.push({
      node: shapePatch === undefined
        ? node
        : { ...node, shape: { ...node.shape!, ...shapePatch } as RigNode["shape"] },
      parents,
    });
    for (const child of node.children ?? []) walk(child, [...parents, node.id]);
  };
  walk(config.rig.root, []);

  const base = config.character.inks;
  const inks = patch?.inks === undefined
    ? base
    : Object.fromEntries(Object.entries(base).map(([name, ink]) => {
        const over = patch.inks![name];
        return [name, over === undefined ? ink : { ...ink, ...over }];
      }));

  return { config, flat, inks };
}

/** Write every channel's rest value into the store. The loader already derived
 *  the map from the rig tree (Task 11), so there is exactly one definition of
 *  "at rest" and this function cannot drift from it. */
export function seedChannels(config: CharacterConfig, channels: Channels): void {
  for (const [key, value] of config.rest) channels.set(key, value as ChannelValue);
}

const num = (channels: Channels, name: string, fallback: number): number => {
  const v = channels.get(name);
  return typeof v === "number" ? v : fallback;
};

/** Rebuild a node's path. A morphable node's geometry lives on its `.shape`
 *  channel — seeded from the rig, then driven by poses, timelines and morphs —
 *  so the channel is authoritative whenever the node has one. The three
 *  primitives have no `.shape` channel and are rebuilt from the rig each frame,
 *  because `bend` and their own parameters are what animate them. */
function resolvePath(node: RigNode, channels: Channels): string {
  const s = node.shape!;
  const driven = channels.get(`${node.id}.shape`);
  if (typeof driven === "string") return driven;
  switch (s.kind) {
    case "ring":
      return ring(s.cx, s.cy, s.r, s.band);
    case "disc":
      return disc(s.cx, s.cy, s.r);
    case "arc":
      return arc(s.cx, s.cy, s.r, s.from, s.to);
    case "polyline":
      return polyline(s.points);
    case "cubicO":
      return cubicO(s.cx, s.cy, s.rx, s.ry);
    case "bezier": {
      if (s.bend === undefined) return bezier(s.points);
      // `bend` offsets each point along one axis by its weight, damped when the
      // bend goes inward — the exact shape of the original's bend(side, a).
      const a = num(channels, `${node.id}.bend`, 0);
      const scaled = Math.sign(a) === s.bend.inwardSign ? a * s.bend.inwardDamp : a;
      const axis = s.bend.axis === "x" ? 0 : 1;
      const moved = s.points.map((p, i): [number, number] => {
        const out: [number, number] = [p[0], p[1]];
        out[axis] += (s.bend!.weights[i] ?? 0) * scaled;
        return out;
      });
      return bezier(moved);
    }
    default:
      throw new Error(`unknown shape kind: ${(s as { kind: string }).kind}`);
  }
}

/** ink key | palette key | "@nodeId" -> a literal colour.
 *
 *  "@body" is late-bound: every mood recolours the whole character by driving one
 *  channel, so the rig never repeats the mood colour on fourteen nodes. The depth
 *  guard is not decoration — a config that made "@x" resolve back to itself would
 *  otherwise hang the render loop rather than fail, and the loader cannot see a
 *  cycle that only closes through a runtime channel value. */
function resolveInk(raw: string, scene: Scene, channels: Channels, depth = 0): string {
  if (depth > 8) throw new Error(`avatar: ink "${raw}" does not resolve to a colour`);
  // A colour channel already holds a literal — the loader normalised it, and a
  // tween mid-flight is producing new literals every frame.
  if (raw.startsWith("#")) return raw;
  if (raw.startsWith("@")) return resolveLateBound(raw, scene, channels, depth);
  const ink = scene.inks[raw];
  if (ink !== undefined) return resolveColourRef(ink.color, scene, channels, depth + 1);
  const colour = scene.config.character.palette[raw];
  if (colour === undefined) throw new Error(`avatar: unknown colour "${raw}"`);
  return colour;
}

/** Resolves an `Ink.color` field. By that field's own contract (see
 *  `config/types.ts`) it is always "#hex", "@nodeId", or a bare PALETTE key —
 *  never another ink key — so, unlike `resolveInk`, this never re-enters the
 *  ink table. That distinction matters for real data: an ink is routinely
 *  named after the very palette entry it points at — `config/load.ts`'s
 *  `paletteColour` comment gives `{ kind: "fill", color: "shell" }` under the
 *  key `shell` as the shape of it — and recursing through the ink table here
 *  would find that same-named ink again and resolve to itself, looping until
 *  the depth guard fires instead of reaching the palette. */
function resolveColourRef(raw: string, scene: Scene, channels: Channels, depth: number): string {
  if (depth > 8) throw new Error(`avatar: ink "${raw}" does not resolve to a colour`);
  if (raw.startsWith("#")) return raw;
  if (raw.startsWith("@")) return resolveLateBound(raw, scene, channels, depth);
  const colour = scene.config.character.palette[raw];
  if (colour === undefined) throw new Error(`avatar: unknown colour "${raw}"`);
  return colour;
}

function resolveLateBound(raw: string, scene: Scene, channels: Channels, depth: number): string {
  const driven = channels.get(`${raw.slice(1)}.ink`);
  if (typeof driven !== "string") {
    throw new Error(`avatar: late-bound ink "${raw}" has no value on ${raw.slice(1)}.ink`);
  }
  return resolveInk(driven, scene, channels, depth + 1);
}

export function compose(scene: Scene, channels: Channels): DisplayList {
  const world = new Map<string, Mat>();
  const out: DisplayItem[] = [];

  for (const { node, parents } of scene.flat) {
    const parentMat = parents.length === 0
      ? IDENTITY
      : world.get(parents[parents.length - 1]!)!;

    const local = fromTransform({
      x: num(channels, `${node.id}.x`, 0),
      y: num(channels, `${node.id}.y`, 0),
      rotation: num(channels, `${node.id}.rotation`, 0),
      scaleX: num(channels, `${node.id}.scaleX`, 1),
      scaleY: num(channels, `${node.id}.scaleY`, 1),
      // Channels, not `node.transform.pivot`: the rig only seeds the rest value.
      // The original moves a transform origin as freely as it moves a rotation —
      // the sad droop turns the face about its bbox bottom, the settle turns it
      // about 60% height — so the origin has to be animatable too.
      pivot: [
        num(channels, `${node.id}.pivotX`, 0),
        num(channels, `${node.id}.pivotY`, 0),
      ],
    });
    const m = multiply(parentMat, local);
    world.set(node.id, m);

    if (!node.shape) continue;

    const d = resolvePath(node, channels);
    const inkChannel = channels.get(`${node.id}.ink`);
    const inkName = typeof inkChannel === "string" ? inkChannel : node.ink!;
    // Stroke-vs-fill and the stroke width are properties of the INK, not of the
    // shape — that is what lets `body` restyle six nodes from one place, and it
    // is the single fact CoreGraphics has to agree with SVG about.
    const ink = scene.inks[inkName];
    out.push({
      id: node.id,
      m,
      d,
      kind: parsePath(d).kind,
      paint: {
        ink: resolveInk(inkName, scene, channels),
        alpha: num(channels, `${node.id}.alpha`, 1),
        fill: ink?.kind === "fill",
        width: ink?.width,
      },
    });
  }
  return out;
}

/** The display list, keeping only what the named crop asks for.
 *
 *  A crop names FEATURES (`character.crops`), and a feature is declared on a
 *  node — but not necessarily on the node that carries the shape. A grouping
 *  node (an eye, say) can declare a feature while the primitives underneath it
 *  — its background, its ring, its interior — declare nothing, because the
 *  feature is a property of the group rather than of each shape it is built
 *  from. So a feature is INHERITED from the nearest ancestor that declares one.
 *
 *  An item whose chain declares none at all is kept by every crop: a node nobody
 *  assigned to a feature is structural, and a crop that silently subtracted
 *  things its author never named would be impossible to reason about.
 *
 *  It has no opinion about visibility either — `compose` emits an `alpha <= 0`
 *  item deliberately, and dropping one here would change what a golden log
 *  records. Deciding an invisible item is not worth painting belongs to the
 *  renderer, and the still-mark renderer (Plan B, Task 39) is where that
 *  decision is made. */
export function cropList(
  config: CharacterConfig,
  list: DisplayList,
  crop: string,
): DisplayList {
  // `hasOwn`, not a plain index-then-`undefined` check: `crops` is a
  // JSON-derived Record, so a plain index sees `Object.prototype` too, and
  // `crops["constructor"]` comes back truthy and not iterable, throwing a
  // `TypeError` instead of this function's own message.
  if (!Object.hasOwn(config.character.crops, crop)) {
    throw new Error(`avatar: unknown crop "${crop}"`);
  }
  const wanted = config.character.crops[crop];
  const features = new Set(wanted);
  const featureOf = new Map<string, string>();

  const walk = (node: RigNode, inherited: string | undefined): void => {
    const feature = node.feature ?? inherited;
    if (feature !== undefined) featureOf.set(node.id, feature);
    for (const child of node.children ?? []) walk(child, feature);
  };
  walk(config.rig.root, undefined);

  return list.filter((item) => {
    const feature = featureOf.get(item.id);
    return feature === undefined || features.has(feature);
  });
}
